### ISSUE TITLE
Bug: [@typescript-eslint/no-invalid-void-type] Complains about `Promise.withResolvers<void>()`

### ISSUE THREAD

**Issue by DesselBane** [typescript-eslint/typescript-eslint#10418]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=LAKAxg9gdgzgLgAgN4IA4CcIFsCWMCmANAuvjBADYBuRJ%2BAVvmIgL4IC8CACprgQHQB3HHAAWAJTKUa6GAB4qEHABMAfAAoAlEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6MAG1w2BZETRoHaJAA08hVgV7IA%2BBwDuAQWQAVABaVkABT7Q%2BAW0T5l0gGYDkiLXv1DEwBJJgBxRCZlSjILEkRTaABzWFcWVAx8OERtbABfXIBdeQK8oA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const { promise, resolve, reject } = Promise.withResolvers<void>()
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-invalid-void-type": [
        "error",
      {
          "allowAsThisParameter": false,
          "allowInGenericTypeArguments": true
      }
    ]
  },
};
```

### tsconfig

```jsonc
{}
```

### Expected Result

No error when specifying the `void` type for the `withResolvers` function.

I want to receive a `Promise<void>` since otherwise its default type is `Promise<unknown>` for which I would need to call `resolve(undefined)` instead of `resolve()`.

### Actual Result

Error: `void is only valid as a return type or generic type argument.`

**Comment by kirkwaiblinger** [comment]
> I want to receive a `Promise<void>` since otherwise its default type is `Promise<unknown>` for which I would need to call `resolve(undefined)` instead of `resolve()`.

Yeah, to be clear this not just a "stylistic" thing about wanting to be able to use `resolve()` instead of `resolve(undefined)`, it's a correctness thing, since it's the _only_ way to have the returned promise be `Promise<void>`, at least with the existing type declarations.

I'm +1 on supporting this, but I'm not quite certain whether this can be generalized, or it would be a special case. The rule, TS's type checking, and TS's type declaration for `Promise.withResolvers()` all have some confusing behavior around this. [Here's what I was exploring](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=PTAEAkHsHdQSVNSBXANgE1ACwIYDcBTUAc2QIGdyDMAXLIgOxxoEtDQAFAJ0gFsWqAOi4VIqQgAoAlKABmyBgGNWkBqAGh0BRahwj0ggLAAoLTr1F5SlWoCyAT258BBAEqjxBAHI42BADwAKgB8Eng4qGQAXKCBUjFO-FRBwQDcJiYgoABCyDSgdBoaOLEAyqAEXDxcoAC0oACMoHqkvAQM%2BQQAHgAO2jTUADQkkPkADEbGiqrk%2BT1NALygDoku7uRihD5%2B0unGWeQ4baCzkFz2JtMMs6A9AEygSys8SW4eW76E-niQLOihUj2mTAgUgJ2YAlk9jKw3sKFAvGQNx6eWaam6qBYihY%2BQABgotLIWAxqLjQFEJIJQMD4AUsMSANZ0orkOlEXijNgQ1SgABGBHpDFo9FAXDQRGJbLkLC4yN0igIkyuyIAzI9lo4Xmt3t5PgEfn9QgSCESSehARl9mBsgBVQKgBijIh0ZjM1kRDbg1jkIkUMqCAOgAD8QZpAcEOCFzVQnroRDFqCI6EgFAYAHJ8iIemcaEHQKUCAReKzMQyiCVechiAUwY78op6IomcTZpHWNz0b0RJQWDNQ1MZvksMgsOqJM9nFR1ptdX5vr9-lJdpasrU1%2BuN5ut9ud7u9yuwFgYFLVkJuzPpHIFMpe2pispkBFUNCzLp9JNXxYr9ZbxrT28Nk8ABBB8ImkBItWSA1-j2T8RG-G8eQnV5p2A0DUBSMIImiWJ4k4SCAhCIErUQM4GXIABCS5B1uAAWdVkO1QDCBAmhH1QZdjBpQJykqapmmIXwGCpLxRilQ5jkKBhiCVGiegAVgYzVJwAmdWPY%2BdDU4mk910vT9N3S0aHsPp8JUgB1HEsFQwhZXcNiuAYAA1bDCOCdUAG8TFAW4CIglSUj2Hzz08GIsMiAgYntAAfMzXgAGRYMsUhkBZ3OgoLRQIAArfowpEHANgYIMYkjexUvShc9gAX0tGkj1gOM4pcQRoCsmzKnIS8rEQu9WTg6gTAGhCbD-AjLLoDrZXUiJMLw-8JusnU7IIBznNclJLWVfIPKy5jIr2mcmmqpSFva5byBmjiLRIniKiqM46kaZouFadpOl6fpqFACRIFkAoTKIBQGUdaAGBkCkaUKVleGYAZZROYkFXUGg01ZK4SWUb6aDBJrjNMv6pRENbvp6AjJhCwgGm0276VZDRoFwfI4WQUAAH1cEINma00SAqW4kUEwlVlZF0YhiGJYhhkrfI4xENGHTBfEhRNYlSTqdzcWgslCvIFhiAYNoOj5AhJd8ghCA6IYTjBRnmAVprs2JXNqOuHbDtCj3CAeE6nmU15Fqmy70M0-5OMpgg7hp1cDNjuPDK4kigPUVl6nII80GFIh-1a879tlbrrxsGkcHQ58Tc0bQ32%2B0tnTp-shqrr8etGxiqEDi64C0HAgnVaCAX8gO85nFa1pciLNsT7bQF2iOYgjtVfbGizh88WUu4IHBOKyO6%2BMe%2BomhaZAjY%2BvoscwX7-vxoGGBBmBwZMBeaen2edXnnV6KXtuCA7-PyA3nu-dw4f2jmABApAKCsjrOGKkeslB13aETVayBHKkwIqRTO2B8BEFxP%2BUOwQyS8hfCaHAaBMzMHoDUF0ahcEEX8MDUGDACFAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6MAG1w2BZADufaE0jyIAXXkBfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false). @typescript-eslint/triage-team , thoughts on whether this is an edge case or something generalizable?

**Comment by JoshuaKGoldberg** [comment]
```ts
// Allowed
class Box<T> { }
new Box<void>();

// Not allowed?
Promise.withResolvers<void>();
```

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=PTAEEEBtIewdwKYBMCwAoAxpAhgZ16AEIwAeAPACoB8oA3qAL7oB2CcRpZAbjAJZJUAFAEoA3OnQhQAORgAXUNmjxkAfnQAFAE4wAtr1wIAdHF5yAFgCUEuGJC4ItubnwEjxaIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6MAG1w2BZADufaE0jyIAXXkBfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

Is this ... a straightforward bug in the rule for not allowing explicit type arguments to function calls, even when [`allowInGenericTypeArguments `](https://typescript-eslint.io/rules/no-invalid-void-type/#allowingenerictypearguments) is enabled? That's my vote, unless I'm horribly confusing something?

**Comment by kirkwaiblinger** [comment]
> ```ts
> // Allowed
> class Box<T> { }
> new Box<void>();
> 
> // Not allowed?
> Promise.withResolvers<void>();
> ```
> 
> [[playground link](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=PTAEEEBtIewdwKYBMCwAoAxpAhgZ16AEIwAeAPACoB8oA3qAL7oB2CcRpZAbjAJZJUAFAEoA3OnQhQAORgAXUNmjxkAfnQAFAE4wAtr1wIAdHF5yAFgCUEuGJC4ItubnwEjxaIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6MAG1w2BZADufaE0jyIAXXkBfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)]
> 
> Is this ... a straightforward bug in the rule for not allowing explicit type arguments to function calls, even when [`allowInGenericTypeArguments`](https://typescript-eslint.io/rules/no-invalid-void-type/#allowingenerictypearguments) is enabled? That's my vote, unless I'm horribly confusing something?

Oh, right, I thought that at first as well, but when I made a rule change, I found out that it's explicitly tested for (with one test case)

https://github.com/typescript-eslint/typescript-eslint/blob/3f02687d964c1e7888e5db62256d56cd72c4f281/packages/eslint-plugin/tests/rules/no-invalid-void-type.test.ts#L195-L204

**Comment by kirkwaiblinger** [comment]
And there's also these docs examples (pay attention to the `logAndReturn<void>(undefined)`):

The following patterns are considered warnings with `{ allowInGenericTypeArguments: false }`:

```ts option='{ "allowInGenericTypeArguments": false }' showPlaygroundButton
logAndReturn<void>(undefined);

let voidPromise: Promise<void> = new Promise<void>(() => {});
let voidMap: Map<string, void> = new Map<string, void>();
```

The following patterns are considered warnings with `{ allowInGenericTypeArguments: ['Ex.Mx.Tx'] }`:

```ts option='{ "allowInGenericTypeArguments": ["Ex.Mx.Tx"] }' showPlaygroundButton
logAndReturn<void>(undefined);

type NotAllowedVoid1 = Mx.Tx<void>;
type NotAllowedVoid2 = Tx<void>;
type NotAllowedVoid3 = Promise<void>;
```

So... I'm not really sure what the intention is of that option tbh.

**Comment by JoshuaKGoldberg** [comment]
cc @G-Rath from 4.5 years ago 😄 - I don't suppose you remember the intent?

**Comment by G-Rath** [comment]
Pretty sure this is a duplicate of #7227 (in which I end up asking myself exactly that same question, of myself... 😅), though I'm on mobile so hard to cross reference the things - either way I'd suggest looking over that issue and checking what it does or doesn't answer

(Happy to have a deeper look myself, just won't be back at my laptop for at least a day or so)
